### PR TITLE
feat(1136): Accept AMS OpenID

### LINKED ISSUE #1136: [Backend] Accept AMS OpenId
**Description:**
_Provide a brief background and justification for this issue_
Backend is currently setup for Login.gov including environment variables and fields in the return payload and these will need to be adjusted in a abstracted way to allow our use of ACF's AMS. 

**Acceptance Criteria:** 
_Create a list of functional outcomes that must be achieved to complete this issue_

- [ ] Authentication and Authorization can pass through successfully using AMS.
- [ ] User's request is redirected back to our frontend
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

- [x] Abstract environment variables, JWT, and payload items to authentication source
- [x] Refactor authentication API functions in tdpservice/users/api for multiple sources
- [x] Ensure all transactions and relevant data events are logged to the server.
- [x] Write new tests for AMS authentication
- [x] Implement new OpenID requests for users who choose AMS
- [x] Run Testing Checklist and confirm all tests pass

**Notes:**
_Add additional useful information, such as related issues and functionality that isn't covered by this specific issue, and other considerations that will be helpful for anyone reading this_ 
- We will likely have to add significant rework to testing.

**Supporting Documentation:**
_Please include any relevant log snippets/files/screen shots_
Please see parent epic.

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_

### PR DESCRIPTION
**Pull request by jorgegonzalez**
### Summary of Changes

This PR adds initial support in the backend for accepting AMS OpenID As an authentication provider.

Closes #1136

_Acceptance criteria as stated in the issue_

### How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d
cd tdrs-backend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d 
```

1. Open `http://localhost:3000/` or `https://tdp-frontend-raft.app.cloud.gov` and sign in using the Login With AMS button.
_IMPORTANT NOTE: given limited acf ams config settings, only one deployed site url (the one above) currently works. we'll inquire if this can change but for now, this will NOT work in staging env_
1. Proceed with the login flow in HHS AMS.
1. Ensure that you are redirected to TDRS
1. Ensure you are able to use the app as normal
1. Log out, and get redirected to the main login page with no errors.
1. Test steps should be captured in the demo GIF(s) and/or screenshots below.

> *Demo GIF(s) and screenshots for testing procedure*

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

+ [x] Look up the acceptance criteria in the related issue; paste ACs below in checklist format.
+ [x] Check against the criteria:
 + [x] Authentication and Authorization can pass through successfully using AMS.
 + [x] User's request is redirected back to our frontend
 + [x] Testing Checklist has been run and all tests pass
_also confirmed that external user route still works_
 + [x] README is updated, if necessary
 + [x] `https://tdp-backend-raft.app.cloud.gov/admin/` for logged out user re-routes to frontend (addressing #1520 behavior)
 + [x] @ADPennington reviewed and tested

As Product Owner, @lfrohlich will decide if ACs are met.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)
  + [x] Backend coverage: ___ (see https://github.com/raft-tech/TANF-app README coverage dashboard)

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Does this PR change any linting or CI settings?

### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*
+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with @iamjolly and @ttran-hub using Accessibility Insights reveal any errors introduced in this PR?
    + [See the full Accessibility Assessment plan here.](https://github.com/HHS/TANF-app/blob/main/docs/Technical-Documentation/how-government-will-test-a11y.md)
_n/a: this PR will redirect users to ACF AMS for authentication. ACF AMS meets compliance standards._ 

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful deployment by assigning a 'Deploy with CircleCI - {env}' label*

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?
_yes, BUT, as noted above, this only works on the raft deployed site, given configuration limitations beyond TDP team's control._
cc: @lfrohlich @abottoms-coder 
### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

+ [x] If this PR introduces backend code, is that code documented both inline and overall?
+ [ ] If this PR introduces frontend code, is that code documented both inline and overall?
+ [x] If this PR introduces dependencies, are their licenses documented?
_https://pypi.org/project/requests-mock/_

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any security issues?
_no new issues detected_ 

### Deliverable 8: Context

> * Performance Standard(s): Code must be understandable and contextualized for the reviewers possess the knowledge and background necessary for analysis and constructive criticism to take place.

> * Acceptable Quality Level: Code submitted in the pull request has context.*
+ [x] Does this pull request contain sufficient inline comments providing relevant context for code snippets?
+ [x] Is the code understandable and lucid?
+ [x] Does this pull request provide background for why coding decisions were made?
+ [x] Can you as a reviewer explain and take ownership of these elements presented in this code review?

### CODE REVIEW

File: tdrs-backend/tdpservice/urls.py
Comment (jorgegonzalez) [inline tdrs-backend/tdpservice/urls.py:27]: re: update redirect in Login.gov

**Comment by jorgegonzalez** [comment]
**The following tasks have been completed as of this writing:**

- Abstract environment variables, JWT, and payload items to authentication source
	- New env variables and and payload items added in `common.py` and utilized in `login.py` and `login_redirect_oidc.py` to support AMS
- Refactor authentication API functions in tdpservice/users/api for multiple sources
	-`urls.py`, `login.py.` `utils.py`, and `login_redirect_oidc.py` have been refactored to support AMS sources and payloads, and have been split up from similar Login.gov classes for improved readability and easier testing.
- Implement new OpenID requests for users who choose AMS
	- The frontend has been tweaked to redirect the user's login request to our AMS handler, which creates the outgoing OpernID request.

**Pending Tasks:**

-  Ensure all transactions and relevant data events are logged to the server.
	- We are currently debugging the outgoing login request to AMS which is returning an error. There may still be some configuration or minor changes needed to get the transaction to complete successfully. Once this is done the login event can properly be logged to our server.
-  Write new tests for AMS authentication
	- More tests new to be written to meet our target coverage and several existing tests need to be fixed.

File: tdrs-backend/tdpservice/users/api/utils.py
Comment (jtwillis92) [inline tdrs-backend/tdpservice/users/api/utils.py:95]: Would suggest abstracting this try/except block

**Comment by jorgegonzalez** [comment]
Yesterday Andrew and I made major progress and he was able to complete the authentication flow and log in to TDP through ACF AMS.

**Completed Task:**

- [x]  Ensure all transactions and relevant data events are logged to the server.
	- After much work debugging the outgoing token request and decoding the incoming payloads and such, we were able to access Andrew's user information from the returned access and ID tokens, and used the related email to authenticate Andrew's user into TDP in the Raft env. These events and queries are being appropriately logged, and this integration is effectively complete.

**Pending Tasks:**

-  [ ] Write new tests for AMS authentication
	- More tests need to be written to meet our target coverage and several existing tests still need to be fixed.
	
A standing decision point, however, remains with the `CustomAuthentication` handler and the assumed usage of the `hhs_id`. This remains in the code, while being currently overridden to use the email from the `/userinfo` endpoint, as it currently does *not* return an `hhs_id`. We are unclear how to request this, and I've reached out to Leonid for potential configuration help. However for the purposes of this pull request, the usage of `username` and refactoring out the `hhs_id` logic may be enough to satisfy the ACs until we have further guidance on `hhs_id`, which, in my opinion, is a well-suited additional task for #1170 anyway.

Thoughts @abottoms-coder @ADPennington?

**Comment by abottoms-coder** [comment]
> 

I'd like to get `hhs_id` in the code but for now we can start writing tests and such. If we don't get traction on this point by Monday/Tuesday, then we can consider putting this off until #1170 .

**Comment by jorgegonzalez** [comment]
To clarify it is still in the code now, but the `username` is being leveraged for the auth method instead `            auth_options["username"] = user_info["email"]`

**Comment by jorgegonzalez** [comment]
Yesterday we refactored all existing login.gov-based auth tests to pass with the new respective OIDC classes. We are now working on abstracting the tests for other auth handlers.

**Comment by jorgegonzalez** [comment]
I spoke with Leonid; he explained that their SSO is currently talking to another IDP that then talks to AMS, and that middleman IDP does not pass the `hhs_id` through. Therefore we will not be able to authenticate with the `hhs_id` until we make the switch to using AMS directly (which Leonid suggested could be after February). So for the time being we will continue using the `email` received from AMS for auth, although I'd like to keep the `hhs` logic and stuff in place to be ready for the switch @abottoms-coder . The only lingering risk of using `email` is the rare case of a user's email changing externally in AMS and not being able to auth into TDRS. Given our current user count of 0, I don't think this is much of a concern at the moment.

**Comment by jorgegonzalez** [comment]
```
 codecov/patch — 96.40% of diff hit (target 97.61%) 
```

Just need to bump the coverage a bit more now.

File: .github/workflows/deploy-on-label.yml
Comment (abottoms-coder) [inline .github/workflows/deploy-on-label.yml:63]: Was this just for testing in the raft environment to prevent disruptions? Should this be commented back in?

File: tdrs-backend/docs/remote-development.md
Comment (abottoms-coder) [inline tdrs-backend/docs/remote-development.md:11]: I think we need to explicitly call out use of v7 here.

File: tdrs-backend/docs/remote-development.md
Comment (abottoms-coder) [inline tdrs-backend/docs/remote-development.md:1]: Thank you for following up on this! This is excellent but could you move it under "Technical Documentation" directory?

File: tdrs-backend/docs/remote-development.md
Comment (abottoms-coder) [inline tdrs-backend/docs/remote-development.md:37]:
```suggestion
Host <app-name>
```

File: tdrs-backend/tdpservice/users/api/login.py
Comment (abottoms-coder) [inline tdrs-backend/tdpservice/users/api/login.py:37]: Is this just for easier traceability for exceptions or did you want to extend functionality here in the future?

File: tdrs-backend/tdpservice/users/api/login.py
Comment (abottoms-coder) [inline tdrs-backend/tdpservice/users/api/login.py:118]:
```suggestion
```

File: tdrs-backend/tdpservice/users/api/login.py
Comment (abottoms-coder) [inline tdrs-backend/tdpservice/users/api/login.py:334]: Comparing this function to the loginDotGov one, should we either explain why headers are necessary only here or also add this there too? Not sure if it's worth it to you to deploy new code just for testing login.gov flows with new headers when it is already functional.

File: tdrs-backend/tdpservice/users/api/login.py
Comment (abottoms-coder) [inline tdrs-backend/tdpservice/users/api/login.py:350]: Could you expand this TODO description to let people know we are blocked by AMS currently? You could add a link to your comment [here](https://github.com/raft-tech/TANF-app/issues/1136#issuecomment-996822564).

File: tdrs-backend/tdpservice/users/api/login_redirect_oidc.py
Comment (abottoms-coder) [inline tdrs-backend/tdpservice/users/api/login_redirect_oidc.py:14]: Is it possible to standardize this with our past logging env var as shown [here](https://github.com/raft-tech/TANF-app/pull/1457/files#diff-15c50d3b3787aa9aa9caa3e31080db56148912a3ac2e1f23e39f3f884bbe2b7c)? I recall we ran into an issue when we were doing remote development so we discarded it then to make progress on this ticket.

```suggestion
logger = logging.getLogger(__name__)

```

File: tdrs-backend/tdpservice/users/api/login_redirect_oidc.py
Comment (abottoms-coder) [inline tdrs-backend/tdpservice/users/api/login_redirect_oidc.py:60]: Great inline comment

File: tdrs-backend/tdpservice/users/api/utils.py
Comment (abottoms-coder) [inline tdrs-backend/tdpservice/users/api/utils.py:41]:
```suggestion
    """ Gets the validation keys to confirm generated nonce and state are correct values. """
```

File: tdrs-backend/tdpservice/users/api/utils.py
Comment (abottoms-coder) [inline tdrs-backend/tdpservice/users/api/utils.py:56]: Should these be re-added back in or are they inaccurate now?

**Comment by jorgegonzalez** [comment]
Test coverage is where it should be now.

File: .github/workflows/deploy-on-label.yml
Comment (jorgegonzalez) [reply .github/workflows/deploy-on-label.yml:63]: Correct, good catch

File: tdrs-backend/tdpservice/users/models.py
Comment (jorgegonzalez) [inline tdrs-backend/tdpservice/users/models.py:25]: Not unused?

File: tdrs-backend/tdpservice/settings/common.py
Comment (jorgegonzalez) [inline tdrs-backend/tdpservice/settings/common.py:386]: Remove

File: tdrs-backend/tdpservice/settings/common.py
Comment (jorgegonzalez) [inline tdrs-backend/tdpservice/settings/common.py:373]: Use `get_required_env_var_setting`

**Comment by jorgegonzalez** [comment]
Fix log out button?

**Comment by jorgegonzalez** [comment]
We need to ask the ops team to update the custom logout redirect url in their configuration. Other than that setting, the current code should be good and this is not a blocker for this ticket. 1170 will be updated to account for this.

**Comment by jorgegonzalez** [comment]
> We need to ask the ops team to update the custom logout redirect url in their configuration. Other than that setting, the current code should be good and this is not a blocker for this ticket. 1170 will be updated to account for this.

This has been addressed and is working as expected now.

File: tdrs-backend/tdpservice/settings/common.py
Comment (jorgegonzalez) [inline tdrs-backend/tdpservice/settings/common.py:113]: Updated this to the `FRONTEND_BASE_URL` per @raftmsohani to account for #1520

File: tdrs-backend/tdpservice/users/api/login.py
Comment (jorgegonzalez) [reply tdrs-backend/tdpservice/users/api/login.py:37]: For easier traceability for exceptions and improved test coverage

File: codecov.yml
Comment (jorgegonzalez) [inline codecov.yml:19]: Patch diff status runs turned off as they are not useful compared to the main project build.

File: docs/Technical-Documentation/remote-development.md
Comment (jorgegonzalez) [inline docs/Technical-Documentation/remote-development.md:1]: Added remote-development documentation covering development on remotely hosted machines and containers.

File: tdrs-backend/Pipfile
Comment (jorgegonzalez) [inline tdrs-backend/Pipfile:49]: Added for simpler request mocking in tests.

File: tdrs-backend/README.md
Comment (jorgegonzalez) [inline tdrs-backend/README.md:198]: Link to the document above

File: tdrs-backend/.env.example
Comment (jorgegonzalez) [inline tdrs-backend/.env.example:25]: These are required to be set with at least dummy values when running locally.

File: tdrs-backend/docker-compose.yml
Comment (jorgegonzalez) [inline tdrs-backend/docker-compose.yml:68]: Pass on new env vars into Docker context

File: tdrs-backend/setup.cfg
Comment (jorgegonzalez) [inline tdrs-backend/setup.cfg:39]: Tweaked flake line-length per @abottoms-coder for convenience. and ignoring common.py to reduce noise

File: tdrs-backend/tdpservice/conftest.py
Comment (jorgegonzalez) [inline tdrs-backend/tdpservice/conftest.py:39]: Lint fixes only in this file

File: tdrs-backend/tdpservice/users/admin.py
Comment (jorgegonzalez) [inline tdrs-backend/tdpservice/users/admin.py:17]: Add new `hhs_id` to User serializer

File: tdrs-backend/tdpservice/users/authentication.py
Comment (jorgegonzalez) [inline tdrs-backend/tdpservice/users/authentication.py:22]: Refactored and added support for future `hhs_id` use

File: tdrs-backend/tdpservice/users/migrations/0027_user_hhs_id.py
Comment (jorgegonzalez) [inline tdrs-backend/tdpservice/users/migrations/0027_user_hhs_id.py:1]: Migration related to new field in User model

File: tdrs-backend/tdpservice/users/models.py
Comment (jorgegonzalez) [inline tdrs-backend/tdpservice/users/models.py:29]: As mentioned above, this isn't currently being used but it is ready to be in the future.

File: tdrs-backend/tdpservice/users/test/test_utils.py
Comment (jorgegonzalez) [inline tdrs-backend/tdpservice/users/test/test_utils.py:58]: Add additional test for missing state tracker is session

File: tdrs-backend/tdpservice/settings/common.py
Comment (jorgegonzalez) [inline tdrs-backend/tdpservice/settings/common.py:370]: Not sure if this should be kept here or reserved to .env vars, as this diverges from login.gov and its default values above.

File: tdrs-backend/tdpservice/urls.py
Comment (jorgegonzalez) [inline tdrs-backend/tdpservice/urls.py:31]: New Authorization and Redirect urls and respective Views

File: tdrs-backend/tdpservice/users/api/login.py
Comment (jorgegonzalez) [inline tdrs-backend/tdpservice/users/api/login.py:49]: This is now a parent class with abstract methods implemented by the respective AMS and LogivDotGov authorization classes below

File: tdrs-backend/tdpservice/users/api/login.py
Comment (jorgegonzalez) [inline tdrs-backend/tdpservice/users/api/login.py:255]: Each `TokenAuthorizationOIDC` subclass will implement `decode_payload`, `get_token_endpoint_response`, and `get_auth_options`

**Review by abottoms-coder — APPROVED** [review]

File: tdrs-backend/.env.example
Comment (ADPennington) [reply tdrs-backend/.env.example:25]: per discussion with Jorge--clarified that these need to be set to the correct values locally.

File: tdrs-backend/tdpservice/settings/common.py
Comment (jorgegonzalez) [inline tdrs-backend/tdpservice/settings/common.py:370]:
```suggestion
        'https://sso-stage.acf.hhs.gov/auth/realms/ACF-SSO/.well-known/openid-configuration'
```

File: tdrs-backend/tdpservice/settings/common.py
Comment (ADPennington) [reply tdrs-backend/tdpservice/settings/common.py:113]: confirming that `https://tdp-backend-qasp.app.cloud.gov/admin/` redirects me to frontend login, as we'd hoped.

**Review by ADPennington — APPROVED** [review]
BIG MILESTONE! 🚀 thank you @jorgegonzalez 

Some important notes:
- i updated ams env var/val pairs in cloud.gov for qasp and staging envs to complete the review for this work. 
- noted below that I had to set accurate the env values to test locally. we should have a note about this saved somewhere centrally for future ref. 
- we're limited to one url right now for the configuration, so as of today, this will only work on the raft deployed site (`https://tdp-frontend-raft.app.cloud.gov/welcome`) @lfrohlich 

ready to merge @abottoms-coder

File: tdrs-backend/tdpservice/users/api/utils.py
Comment (jorgegonzalez) [reply tdrs-backend/tdpservice/users/api/utils.py:56]: This is inaccurate now

**Comment by abottoms-coder** [comment]
> BIG MILESTONE! 🚀 thank you @jorgegonzalez
> 
> Some important notes:
> 
> * i updated ams env var/val pairs in cloud.gov for qasp and staging envs to complete the review for this work.
> * noted below that I had to set accurate the env values to test locally. we should have a note about this saved somewhere centrally for future ref.
> * we're limited to one url right now for the configuration, so as of today, this will only work on the raft deployed site (`https://tdp-frontend-raft.app.cloud.gov/`) @lfrohlich
> 
> ready to merge @abottoms-coder

@jorgegonzalez Do we have a follow-on task for updating AMS to allow `-staging` or were we going to wait for the prod URL?

**Comment by jorgegonzalez** [comment]
> @jorgegonzalez Do we have a follow-on task for updating AMS to allow -staging or were we going to wait for the prod URL?

Not that I know of, I defer to you and @ADPennington for the appropriate response there

**Comment by ADPennington** [comment]
> > @jorgegonzalez Do we have a follow-on task for updating AMS to allow -staging or were we going to wait for the prod URL?
> 
> Not that I know of, I defer to you and @ADPennington for the appropriate response there

i think there is a feasibility question here for ACF AMS Ops. We can track that discussion in a spike ticket that links to the acf ams epic. @abottoms-coder @jorgegonzalez @valcollignon

**Comment by abottoms-coder** [comment]
> > > @jorgegonzalez Do we have a follow-on task for updating AMS to allow -staging or were we going to wait for the prod URL?
> > 
> > 
> > Not that I know of, I defer to you and @ADPennington for the appropriate response there
> 
> i think there is a feasibility question here for ACF AMS Ops. We can track that discussion in a spike ticket that links to the acf ams epic. @abottoms-coder @jorgegonzalez @valcollignon

This is now to be tracked in #1552

**Pull request merged**
